src/HydraAuction/Hydra/Monad.hs

```diff
@@ -0,0 +1,93 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module HydraAuction.Hydra.Monad (
+  MonadHydra (..),
+  EventMatcher (..),
+  AwaitedHydraEvent (..),
+  waitForHydraEvent,
+  sendCommandAndWaitFor,
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Haskell imports
+import Control.Monad (void)
+import GHC.Natural (Natural)
+
+-- Cardano imports
+import Cardano.Api (AddressInEra (..))
+import Cardano.Api.UTxO qualified as UTxO
+
+-- Hydra imports
+import Hydra.Cardano.Api (Address (ByronAddress, ShelleyAddress), Tx)
+import Hydra.Cardano.Api.Prelude (TxOut (..))
+
+-- HydraAuction imports
+import HydraAuction.Hydra.Interface (
+  HydraCommand (GetUTxO, NewTx),
+  HydraEvent (GetUTxOResponse, SnapshotConfirmed),
+  HydraEventKind (GetUTxOResponseKind),
+ )
+import HydraAuctionUtils.Monads (
+  MonadQueryUtxo (..),
+  MonadSubmitTx (..),
+  UtxoQuery (..),
+ )
+
+data AwaitedHydraEvent
+  = Any
+  | SpecificEvent HydraEvent
+  | SpecificKind HydraEventKind
+  | CustomMatcher EventMatcher
+  deriving stock (Show)
+
+newtype EventMatcher = EventMatcher (HydraEvent -> Bool)
+
+instance Show EventMatcher where
+  show (EventMatcher _) = "EventMatcher <some HydraEvent predicate>"
+
+class Monad m => MonadHydra m where
+  sendCommand :: HydraCommand -> m ()
+  waitForHydraEvent' :: Natural -> AwaitedHydraEvent -> m HydraEvent
+
+defaultTimeout :: Natural
+defaultTimeout = 30
+
+waitForHydraEvent :: MonadHydra m => AwaitedHydraEvent -> m HydraEvent
+waitForHydraEvent = waitForHydraEvent' defaultTimeout
+
+sendCommandAndWaitFor :: MonadHydra m => AwaitedHydraEvent -> HydraCommand -> m HydraEvent
+sendCommandAndWaitFor awaitedSpec command = do
+  sendCommand command
+  waitForHydraEvent awaitedSpec
+
+-- FIXME: this does not work without `Monad m`, do not know why
+instance {-# OVERLAPPABLE #-} (Monad m, MonadHydra m) => MonadSubmitTx m where
```

Comment:
I read the discussion in the other comment, I agree we can defer this to a follow-up PR. Maybe we could just add a comment here explaining why we are using overlappable

Reply:
Follow-up #237